pandas/core/sorting.py

```diff
@@ -285,6 +299,186 @@ def nargsort(
     return indexer
 
 
+def ensure_key_mapped_dataframe(df, key: Callable, levels=None, axis=0):
+    """
+    Returns a new DataFrame in which key has been applied
+    to all levels specified in level (or all levels if level
+    is None) along an axis. Used for key sorting for DataFrames.
+
+    Parameters
+    ----------
+    df : DataFrame
+        DataFrame to which to apply the key function on the
+        specified levels.
+    key : Callable
+        Function that takes a Series and returns a Series of
+        the same shape. This key is applied to each level separately.
+    levels : list-like, int or str, default None
+        Level or list of levels to apply the key function to.
+        If None, key function is applied to all levels. Other
+        levels are left unchanged.
+    axis : int, default 0
+        Axis to apply keys to. Expects 0 or 1.
+
+    Returns
+    -------
+    df : DataFrame
+        Resulting DataFrame with modified data.
+    """
+    from pandas import DataFrame, Series
+
+    if not isinstance(df, DataFrame):
+        raise TypeError(
+            "ensure_key_mapped_dataframe can only be called on DataFrame objects."
+        )
+
+    if axis == 0:
+        axis_levels = list(df.columns._values)  # makes mypy happy
+    else:
+        axis_levels = list(df.index._values)
+
+    if levels is not None:
+        if isinstance(levels, (str, int)):
+            sort_levels = [levels]
+        else:
+            sort_levels = levels
+    else:
+        sort_levels = axis_levels
+
+    values = [
+        (name, Series(df._get_label_or_level_values(name, axis=axis), name=name))
+        for name in axis_levels
+    ]
+
+    new_levels = [
+        ensure_key_mapped(series, key) if name in sort_levels else series
+        for (name, series) in values
+    ]
+
+    if axis == 0:
+        new_df = DataFrame._from_arrays(new_levels, df.columns, df.index)
+    else:
+        new_df = DataFrame._from_arrays(new_levels, df.index, df.columns).transpose()
```

Comment:
Is there a way to do this (construct a DataFrame along either row or column axis) that's better than transposing it like this?